src/coreclr/gc/gc.cpp

```diff
@@ -49307,3 +49377,294 @@ void PopulateDacVars(GcDacVars *gcDacVars)
+        // to very powerful machines - before that's implemented, temporarily I am only enabling this for 80GB+ memory.
+        // For now I am using an estimate to calculate these numbers but this should really be obtained
+        // programmatically going forward.
+        // I am assuming 47 processes using WKS GC and 3 using SVR GC.
+        // I am assuming 3 in part due to the "very high memory load" is 97%.
+        int available_mem_th = 10;
+        if (total_physical_mem >= ((uint64_t)80 * 1024 * 1024 * 1024))
+        {
+            int adjusted_available_mem_th = 3 + (int)((float)47 / (float)g_num_processors);
+            available_mem_th = min (available_mem_th, adjusted_available_mem_th);
+        }
+
+        high_memory_load_th = 100 - available_mem_th;
+        v_high_memory_load_th = 97;
+    }
+
+    m_high_memory_load_th = min ((high_memory_load_th + 5), v_high_memory_load_th);
+}
+
+void gc_heap::refresh_memory_limit()
+{
+    if (GCConfig::GetGCTotalPhysicalMemory() != 0)
+    {
+        return;
+    }
+
+    GCToEEInterface::SuspendEE(SUSPEND_FOR_GC);
+    decommit_lock.Enter();
+
+    uint32_t nhp_from_config = static_cast<uint32_t>(GCConfig::GetHeapCount());
+#ifdef MULTIPLE_HEAPS
+    uint32_t nhp = n_heaps;
+#else
+    uint32_t nhp = 1;
+#endif //MULTIPLE_HEAPS
+    size_t seg_size_from_config;
+
+    total_physical_mem = GCToOSInterface::GetPhysicalMemoryLimit (&is_restricted_physical_mem);
+
+    bool had_hard_limit = heap_hard_limit != 0;
+#ifdef COMMITTED_BYTES_SHADOW
+    // Stress the committed bytes reconstruction logic by pretending a job object exists
+    // with memory == physical memory limit and pretending a hard limit was not there
+    // The combined result is that we will force the committed bytes reconstruction all the time.
+    is_restricted_physical_mem = true;
+    had_hard_limit = false;
+#endif //COMMITTED_BYTES_SHADOW
+
+    configure_memory_settings(false, nhp, nhp_from_config, seg_size_from_config);
+
+    if (!had_hard_limit && heap_hard_limit)
+    {
+#ifndef USE_REGIONS
+        assert (!"NYI - Segments");
+#endif //!USE_REGIONS
+        check_commit_cs.Initialize();
+
+        // Accounting for the bytes committed for the regions
+        size_t total_committed = 0;
+        size_t committed_bookkeeping = 0;
+#ifdef USE_REGIONS
+        for (int oh = soh; oh < total_oh_count; oh++)
+        {
+            int s = (oh == 0) ? 0 : oh + 2;
+            int e = oh + 2;
+            size_t total_committed_per_oh = 0;
+#ifdef MULTIPLE_HEAPS
+            for (int h = 0; h < n_heaps; h++)
+            {
+                gc_heap* heap = g_heaps [h];
+#else
+            {
+                gc_heap* heap = nullptr;
+#endif //MULTIPLE_HEAPS
+                size_t total_committed_per_heap = 0;
+                for (int gen = s; gen <= e; gen++)
+                {
+                    heap_segment* seg = generation_start_segment (heap->generation_of (gen));
```

Comment:
you could use the conventional way to do this in GC which is to start from the first RW region -

`heap_segment_rw (generation_start_segment (generation_of (gen)));`

so you don't need the heap_segment_read_only_p check later